apps/examples/smartfs/buffer_optimization/buffer_optimization_main.c

```diff
@@ -0,0 +1,466 @@
+
+	unlink(SAMPLE_FILE);
+
+	ret = create_test_file(size * 1024);
+	if (ret != OK) {
+		printf("Unable to complete test\n");
+		return -1;
+	}
+
+	ret = stat("/mnt", &st);
+	if (ret < 0) {
+		printf("Stat to get optimal read buffer size failed, Errno = %d\n", errno);
+		return -1;
+	}
+	buf_size = st.st_blksize;
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_1) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	ret = append_test(buf_size);
+	if (ret != OK) {
+		printf("Appending with buffer size %d failed\n", buf_size);
+		return ret;
+	}
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_2) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	time_append = time_2.timeleft - time_1.timeleft;
+	printf("TEST WITH OPTIMIZED BUFFER\n");
+	print_test_result("appending", TOTAL_APPEND_SIZE, APPEND_TEST_ITR, buf_size, time_append);
+
+	unlink(SAMPLE_FILE);
+
+	printf("================================= APPEND TESTS COMPLETE =================================\n\n");
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: read_test
+ *
+ * Description: Perform a back-to-back read of the entire file's contents,
+ *              multiple times.
+ *
+ ****************************************************************************/
+
+static int read_test(int size)
+{
+	int fd;
+	int ret;
+	int itr = 0;
+
+	while (itr < READ_TEST_ITR) {
+		fd = open(SAMPLE_FILE, O_RDOK);
+		if (fd < 0) {
+			printf("Unable to open Sample File while testing, Errno = %d\n", errno);
+			return ERROR;
+		}
+
+		do {
+			ret = read(fd, g_buf, size);
+		} while (ret == size);
+
+		close(fd);
+		itr += 1;
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: read_test_util
+ *
+ * Description: Set up the timer and run the read test with different
+ *              buffer sizes.
+ *
+ ****************************************************************************/
+
+static int read_test_util(int frt_fd, int size)
+{
+	int ret;
+	int buf_size;
+	struct timer_status_s time_1, time_2;
+	uint32_t time_read;
+	struct stat st;
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_1) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	ret = read_test(512);
+	if (ret != OK) {
+		printf("Reading with buffer size 512 failed\n");
+		return ret;
+	}
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_2) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	time_read = time_2.timeleft - time_1.timeleft;
+	print_test_result("reading", size, READ_TEST_ITR, 512, time_read);
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_1) < 0) {
```

Comment:
ioctl will return one of error value that is defined in errno.h, so you'd better use below style
```
ret = ioctl();
if (ret < 0) {
    printf();
    return ret;
}
```